numba/core/typing/asnumbatype.py

```diff
@@ -0,0 +1,193 @@
+import inspect
+import typing as pt
+import typing_extensions as pt_ext
+import numpy as np
+
+from numba.core.typing.typeof import typeof
+from numba.core import errors, types, utils
+from numba.np import numpy_support
+
+
+def _py_version_switch(py_version, new_result, old_result):
+    return new_result if utils.PYVERSION >= py_version else old_result
+
+
+class AsNumbaTypeRegistry:
+    """
+    A registry for python typing declarations.  This registry stores a lookup
+    table for simple cases (e.g. int) and a list of functions for more
+    complicated cases (e.g. generics like List[int]).
+
+    The as_numba_type registry is meant to work statically on type annotations
+    at compile type, not dynamically on instances at runtime. To check the type
+    of an object at runtime, see numba.typeof.
+    """
+
+    def __init__(self):
+        self.lookup = {
+            type(example): typeof(example)
+            for example in [
+                0,
+                0.0,
+                complex(0),
+                "numba",
+                True,
+                None,
+            ]
+        }
+
+        self.functions = [
+            self._builtin_infer, self._numba_type_infer, self._numpy_type_infer,
+        ]
+
+    def _numba_type_infer(self, py_type):
+        if isinstance(py_type, types.Type):
+            return py_type
+
+    def _numpy_type_infer(self, py_type):
+        if inspect.isclass(py_type) and issubclass(py_type, np.generic):
+            return numpy_support.from_dtype(py_type)
+
+    def _builtin_infer(self, py_type):
+        # The type hierarchy of python typing library changes in 3.7.
+        generic_type_check = _py_version_switch(
+            (3, 7),
+            lambda x: isinstance(x, pt._GenericAlias),
+            lambda _: True,
+        )
+        if not generic_type_check(py_type):
+            return
+
+        annotated_type_check = _py_version_switch(
+            (3, 7),
+            lambda x: hasattr(x, "__metadata__"),
+            lambda x: getattr(x, "__origin__", None) is pt_ext.Annotated
+        )
+
+        list_origin = _py_version_switch((3, 7), list, pt.List)
+        dict_origin = _py_version_switch((3, 7), dict, pt.Dict)
+        set_origin = _py_version_switch((3, 7), set, pt.Set)
+        tuple_origin = _py_version_switch((3, 7), tuple, pt.Tuple)
+
+        if annotated_type_check(py_type):
+            inner_py_type = py_type.__args__[0]
+            annotations = _py_version_switch(
+                (3, 7),
+                lambda: py_type.__metadata__,
+                lambda: py_type.__args__[1]
+            )()
+
+            if len(annotations) == 1:
```

Comment:
The spec for annotated says you should ignore any arguments you don't recognize, as they may be for other tools. So I'd recommend you filter the annotations to those where `ininstance(arg, numba.Type)` is true, and assert that is of length 1.

Reply:
The [PEP](https://www.python.org/dev/peps/pep-0593/#consuming-annotations) also says
> Multiple annotations: It's up to the tool consuming the annotations to decide whether the client is allowed to have several annotations on one type and how to merge those annotations.

Checking `isinstance(arg, numba.Type)` would disallow the current recursive call where `Annotated[py_type, x] => as_numba_type(x)`.

I'll leave it up to the core devs.  Either we only support explicit numba types, in which case we can do such a check, or we also support numpy and builtin types (e.g. `Annotated[np.ndarray, int, 1]`), in which case I think it's simpler to say that our tool does not support multiple annotations. 

FWIW I wouldn't expect numba to "just work" with other python tools, so supporting their annotations may be misleading.  I don't have much experience with what other tools would use `Annotated` for.

Reply:
I don't think I am familiar with the type annotation standard or convention to make the call.